### PR TITLE
Bug 2134786: Verify that no recovery occurs when norecover flag is set on the cluster

### PR DESCRIPTION
**Pull request by SrinivasaBharath**
# Description

As part of the automation of the verification of steps of the BZ#2134786 following scenarios are automated-

Scenario1: Verify that the pg_num should not  increase after setting the norecover flag
Scenario2: Verify that the pg_num should not decrease after setting the norecover flag
Scenario3:Modify the pg number manually and verify that the pg_num should not increase after setting the norecover flag
Scenario4:Modify the pg number manually and verify that the pg_num should not decrease after setting the norecover flag
Scenario5:Checking the recovery of cluster after setting norecover flag

Bug link- https://bugzilla.redhat.com/show_bug.cgi?id=2134786 

Polarion test case added- https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83591783

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/rados/test_norecovery_flag_functionality.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+"""
+This file contains the  methods to verify the  norecover flag.
```
Comment (pdhiran) [inline tests/rados/test_norecovery_flag_functionality.py:2]: Please add the bugzilla IDs for which the module is written to test.

File: tests/rados/test_norecovery_flag_functionality.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    try:
+        utils.configure_osd_flag(ceph_cluster, "set", "noscrub")
+        log.info("The noscrub is set on the cluster")
+        utils.configure_osd_flag(ceph_cluster, "set", "nodeep-scrub")
+        log.info("The nodeep-scrub is set on the cluster")
+        utils.configure_osd_flag(ceph_cluster, "set", "nobackfill")
+        log.info("The nobackfill is set on the cluster")
+        utils.configure_osd_flag(ceph_cluster, "set", "norecover")
+        log.info("The norecover is set on the cluster")
+        if not rados_obj.create_pool(name=pool_name, **replicated_config):
```
Comment (pdhiran) [inline tests/rados/test_norecovery_flag_functionality.py:43]: Please write some data into the pool once created.

File: tests/rados/test_norecovery_flag_functionality.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            log.error("Failed to create the  Pool")
+            return 1
+        log.info(f"The pool-{pool_name} is created")
+        # check the ceph df
+        end_time = datetime.now() + timedelta(seconds=3600)
```
Comment (pdhiran) [inline tests/rados/test_norecovery_flag_functionality.py:48]: Before entering the loop, please trigger some autoscaler activity.

Please refer to the bug, to get the area where the fix was provided.

**Review by pdhiran — CHANGES_REQUESTED** [review]
The code does not completely verify the fix provided with the bugzilla.

Workflow needs to be updated to verify the bug fix.

File: tests/rados/test_norecovery_flag_functionality.py (hunk 1 above)
Comment (SrinivasaBharath) [reply tests/rados/test_norecovery_flag_functionality.py]: Added the details.

File: tests/rados/test_norecovery_flag_functionality.py (hunk 2 above)
Comment (SrinivasaBharath) [reply tests/rados/test_norecovery_flag_functionality.py]: Writing the data into the pools.

File: tests/rados/test_norecovery_flag_functionality.py (hunk 3 above)
Comment (SrinivasaBharath) [reply tests/rados/test_norecovery_flag_functionality.py]: The following five scenarios are covered-
Scenario1: Verify that the pg_num should not  increase after setting the norecover flag
Scenario2: Verify that the pg_num should not decrease after setting the norecover flag
Scenario3:Modify the pg number manually and verify that the pg_num should not increase after setting the norecover flag
Scenario4:Modify the pg number manually and verify that the pg_num should not decrease after setting the norecover flag
Scenario5:Checking the recovery of cluster after setting norecover flag

**Comment by SrinivasaBharath** [comment]
Log - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-RWQV3X

File: suites/reef/rados/tier-2_rados_test-bugfixes.yaml [hunk 4]
```diff
@@ -168,3 +168,15 @@ tests:
         slow-osd-heartbeat: true
       polarion-id: CEPH-83590688
       desc: Generate Slow OSD heartbeats by inducing network delay
+
```
Comment (pdhiran) [inline suites/reef/rados/tier-2_rados_test-bugfixes.yaml:171]: Is the test fix added only in reef?  If that's the case, let's check if a backport bug needs to be created from QA side.

File: tests/rados/test_norecovery_flag_functionality.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        log.info(e)
+        log.info(traceback.format_exc())
+        return 1
+    finally:
+        log.info("Execution of finally block")
+        if config.get("delete_pool"):
+            method_should_succeed(rados_obj.delete_pool, pool_name)
+            log.info(f"deleted the {pool_name} pool successfully")
+        utils.configure_osd_flag(ceph_cluster, "unset", "norecover")
+        log.info("The norecover is unset on the cluster")
```
Comment (pdhiran) [inline tests/rados/test_norecovery_flag_functionality.py:260]: Please log cluster health in finally.
        rados_obj.log_cluster_health()

Comment (SrinivasaBharath) [reply tests/rados/test_norecovery_flag_functionality.py:267]: Added the code.

File: tests/rados/test_norecovery_flag_functionality.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+    Args:
+        pool_object: pool object
+        current_pg_num: Pg number
+        pool_name: pool name
+
+    Returns: True -> PG number is not changed
+             False -> PG number is changed
+
+    """
+    endtime = datetime.datetime.now() + datetime.timedelta(seconds=180)
```
[the remark this answers is not in the document]
Comment (SrinivasaBharath) [reply tests/rados/test_norecovery_flag_functionality.py]: Increased the time for 5 min

**Comment by SrinivasaBharath** [comment]
Log: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-7RAHQE Recovery

File: suites/reef/rados/tier-2_rados_test-bugfixes.yaml (hunk 4 above)
Comment (SrinivasaBharath) [reply suites/reef/rados/tier-2_rados_test-bugfixes.yaml:171]: Backport bug is raised for Quincy- https://bugzilla.redhat.com/show_bug.cgi?id=2293908

**Review by pdhiran — APPROVED** [review]
Looks good!

**Pull request merged**